### PR TITLE
Added event listener endpoint

### PR DESCRIPTION
**Pull request by xavi-**
**What**: Added an http endpoint to receive marathon events.  Status change events are then logged out using KayVee which are eventually pushed to Influx.

**Why**: By logging when containers are create, destroyed, crash, etc..., teams can create alerts when these things happen too frequently.

Context: https://clever.atlassian.net/browse/INFRA-774

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -59,144 +62,59 @@ func main() {
...
-	}
-	// add up resource counts from slaves
-	totalCPU := 0.0
-	totalMem := int64(0)
-	for _, slave := range state.Slaves {
-		r := slave.Resources
-		totalCPU += r.CPUs
-		totalMem += r.Mem
+func initPushListeners(wg *sync.WaitGroup) {
+	listener := marathon.NewListener(containerHost, containerPort)
```
Comment (dfuentes) [inline main.go]: The listener here should use the PORT env variable, not the container port.

Actually you should pass both.  You need to use the container port to start the go server, and the PORT env variable to register the callback url

**Comment by dfuentes** [comment]
lgtm

File: marathon/logging.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		"mem":            app.Mem,
+		"disk":           app.Disk,
+		"image":          app.Container.Docker.Image,
+		"version":        app.Version,
+		"tasksStaged":    app.TasksStaged,
+		"tasksHealthy":   app.TasksHealthy,
+		"tasksRunning":   app.TasksRunning,
+		"tasksUnhealthy": app.TasksUnhealthy,
+		"type":           "gauge", // This is to auto load metric into influx
+	}
```
Comment (nathanleiby) [inline marathon/logging.go:54]: By emitting all the stats in a single block, I believe that makes it a bit harder to auto-handle from the Heka side. This is at least true for SignalFx; I'm not 100% about Influx -- can it write multiple values per stat?

If not, We would need to write Heka filters for each one of these, specify which field to pull the value from, and give it a new name so it fits the `<source>.<title>` convention.

What if we instead emitted one stat per line, e.g. `marathon-stats.cpus`, `marathon-stats.mem`, etc... where each one has a `type` and `value` (relevant [infra-docs](https://github.com/Clever/infra-docs/blob/master/monitor/application-metrics.md#example))

Comment (dfuentes) [inline marathon/logging.go:54]: Influx has no problem with multiple stats (see http://grafana.ops.clever.com/#/dashboard/db/marathon-apps?panelId=1&fullscreen&edit as an example).  Good point about it causing issues for signalfx.  I'm not sure we need to clean it up now, but if it's easy why not.

Comment (xavi-) [inline marathon/logging.go:54]: I think this change makes more sense as a separate ticket.  This particular function isn't related to the changes made for container life cycle event logging.

https://clever.atlassian.net/browse/INFRA-932

File: marathon/push_listener.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+func (l *Listener) unsubscribe(marathonHost, callback string) error {
+	marathonURL := url.URL{Scheme: "http", Host: marathonHost, Path: "/v2/eventSubscriptions"}
+	q := marathonURL.Query()
+	q.Set("callbackUrl", callback)
+	marathonURL.RawQuery = q.Encode()
+
+	req, err := http.NewRequest("DELETE", marathonURL.String(), nil)
+	res, err := http.DefaultClient.Do(req)
+	defer res.Body.Close()
```
Comment (dfuentes) [inline marathon/push_listener.go]: I think that this potentially causes a panic in the case that there is an error (as res.Body is not guaranteed to be a valid object in that case).

File: marathon/logging.go (hunk 2 above)
Comment (nathanleiby) [inline marathon/logging.go:54]: :+1:

File: marathon/push_listener.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("Bad status code while subscribing to marathon events: %s", res.Status)
+	}
+
+	var data map[string]interface{}
+	decoder := json.NewDecoder(res.Body)
+	if err := decoder.Decode(&data); err != nil {
+		return err
+	}
```
Comment (dfuentes) [inline marathon/push_listener.go]: Why do we decode here?

Comment (xavi-) [inline marathon/push_listener.go]: Whoops.  Removed in f5f2b91f30f6e6181a0d8523eb9a225b0ace2dee.

File: marathon/push_listener.go (hunk 3 above)
Comment (xavi-) [inline marathon/push_listener.go]: Fixed in c5ac5a0dd6b678ab29a73a87f1e0fa9839e7c7bf.

**Comment by dfuentes** [comment]
Looks good, have you tested in dev yet?

**Comment by xavi-** [comment]
Yeah I tested an earlier version of jian's machine on friday.  I'll go a head and test again.

**Comment by dfuentes** [comment]
I meant on marathon in particular, since the internal/external port stuff
there will behave differently.  I'm not too worried about it though.

File: marathon/logging.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+func normalizeAppID(appID string) string {
+	return strings.Replace(appID[1:], "/", ".", -1)
+}
+
+func LogEvent(event Event) {
+	eventLog := m{
+		"source":    "marathon-event",
+		"title":     event.TaskStatus,
+		"level":     kayvee.Info,
```
Comment (nathanleiby) [inline marathon/logging.go:85]: Might we consider varying the log level depending on the type of event? (e.g. container crashed)

Comment (xavi-) [inline marathon/logging.go:85]: Is the log level used anywhere?  If not, I'm inclined to leave the default as `kayvee.Info`.

Comment (nathanleiby) [inline marathon/logging.go:85]: Not yet consumed by anything.

**Pull request merged**